File: golang/subarray_sum.go
func subarraySum(nums []int, k int) int {
	// sum(i, j) = sum(0, j) - sum(0, i-1)
	// k = sum(i, j) -> sum(0, j) - k = sum(0, i-1) meaning i need to find presum before i
	// think of presum as start point and current sum as end point

	count, sum := 0, 0
	freq := map[int]int{}
	freq[0]++ // need to init zero to count presum = 0

	for i := 0; i < len(nums); i++ {
		sum += nums[i]

		// found presum, valid subarray sum
		if f, ok := freq[sum-k]; ok {
			count += f
		}

		// setup presum
		freq[sum]++
	}

	return count
}

File: typescript/subarray_sum.ts
function subarraySum(nums: number[], k: number): number {
	// need to find cases where sum(i, j) = k but sum(i, j) = sum(0, j) - sum(0, i)
	// store prefix sums sum(0, i) in hash set, check if runningSum - k inside prefix sums

	// needs to be a freq map to count duplicate sums
	let prefixSums = new Map<number, number>()
	prefixSums.set(0, 1) // init 0 bc num by itself is valid

	let runningSum = 0
	let count = 0
	for (let i = 0; i < nums.length; i++) {
		runningSum += nums[i]

		// subarray sum is valid
		if (prefixSums.has(runningSum - k)) {
			count += prefixSums.get(runningSum - k)
		}

		prefixSums.set(runningSum, prefixSums.has(runningSum) ? prefixSums.get(runningSum) + 1 : 1)
	}

	return count
}
